refactor: extracted DataDict.asJSONDictionary()

TizianoCoroneo opened the pull request:
Hey there 👋 

recently I found the need to encode a `DataDict` instance into a JSON object. I found that this functionality is already provided but only for `GraphQLResult` instances instead.

 I refactored the relevant code to also expose a `DataDict.asJSONDictionary()` method that performs the same logic as the one in `GraphQLResult`.

AnthonyMDev reviewed (changes requested):
Thanks for the PR @TizianoCoroneo!

The team discussed this and we'd like to request a change before merging this. The `DataDict` is not really intended for public consumption. It has to be public in order for your generated models to access it and compile properly, but we don't want to encourage people to use it directly.

We think this would be improved by making this accessible directly from `SelectionSet` instead of `DataDict`.

We also try not to cloud the namespace of `SelectionSet` too much because of naming collisions and code completion. I haven't talked to the team about this part yet, but I'm wondering if it would be better to have a global function or static struct for this. Something like `Apollo.JSONConverter.convert(_ selectionSet: ApolloAPI.SelectionSet) -> [String: Any]`.

We definitely want to do this on `SelectionSet`, not `DataDict`, but lets discuss the actual API for this before finalizing and merging.

@calvincestari @BobaFetters, any opinions/suggestions?

calvincestari commented:
> I'm wondering if it would be better to have a global function or static struct for this. Something like `Apollo.JSONConverter.convert(_ selectionSet: ApolloAPI.SelectionSet) -> [String: Any]`.

I hadn't considered the package namespace for this but I'm not opposed to it. I think it could help with clarity on where this feature is available.

BobaFetters commented:
Yea I like the idea of having available as a global like you suggested
`Apollo.JSONConverter.convert(_ selectionSet: ApolloAPI.SelectionSet) -> [String: Any]`

And I think we should use the `JSONConverter` to hold all of our helper functions that handle JSON conversion, maybe we go through in future PRs and deprecate some existing functions and bring it all into one place.

TizianoCoroneo commented:
I updated the PR. Now there is an enum with static methods to perform the conversions like this:
```swift
let jsonObject = Apollo.JSONConvert.selectionSetToJSONObject(myHero)
```

I also added a function to convert a `GraphQLResult` into a JSONObject that internally uses `asJSONDictionary`, just because the sight of one empty enum with one static method kind of hurts.

AnthonyMDev commented:
I've branched off of this PR and made a few changes. New PR is up #380. Thanks for the help @TizianoCoroneo!
